lib/mongodb.js

```diff
@@ -1881,7 +1881,11 @@ MongoDB.prototype.isObjectIDProperty = function(model, prop, value, options) {
   ) {
     return true;
   } else if ('string' === typeof value) {
+    var properties = this._models[model] && this._models[model].properties;
+    if (properties.id && properties.id.isObjectID) return true;
```

Comment:
This condition does not look correct to me. You should be checking whether `prop.mongodb.dataType` is set to `ObjectID` (case-insensitively). Possibly also check if `prop.type` is `string`.

Note that we need to perform the coercion for all properties, not only primary keys. Most notably for foreign keys used by models with a belongsTo relation.

Reply:
@bajtos I have updated the implementation.

When `mongodb: {dataType: 'ObjectID'}` is set, any ObjectID-like id will be stored as an ObjectID, overriding the model-level setting of `strictObjectIDCoercion: true`. The id will be stored as an ObjectID, and it will always be resolved as an ObjectID. Therefore there is no problem when it comes to relations.

Regarding your other point "perform the coercion for all properties". Do we want to do that in this PR?

Reply:
>  The id is stored as an ObjectID, and it will always be resolved as an ObjectID. Therefore there is no problem when it comes to relations.

Could you please verify that claim for the following model definition?

```
const properties = {
    // id is explicitly typed as string
    id: {
     type: string,
     id: true,
    },

    // additional properties, not important
    name: 'string',
}

const settings: {
  // please test both options: idInjection disabled and enabled
  idInjection: false
}

const MyModel = ds.define('MyModel', properties, settings);
```

> Regarding your other point "perform the coercion for all properties". Do we want to do that in this PR?

Feel free to leave it out. I think that your changes in `MongoDB.prototype.isObjectIDProperty` should be enough to fix coercion for all properties, so it should be enough to add few tests to verify. Anyhow, let's focus on fixing primary keys first.

Reply:
Without specifying `mongodb: {dataType: 'ObjectID'}`?

>  I think that your changes in MongoDB.prototype.isObjectIDProperty should be enough to fix coercion for all properties, so it should be enough to add few tests to verify

That method is called only for ids at the moment. If we want to coerce all the properties, we'll have to refactor the code to loop through the properties, in many places. That's why I was wondering if we should do it in this PR.

Reply:
> Without specifying ``mongodb: {dataType: 'ObjectID'}`?

Sorry, I forgot to include that flag.

I think the following definition should be correct:

```js
const properties = {
    // id is explicitly typed as string
    id: {
     type: string,
     id: true,
     mongodb: {dataType: 'ObjectID'}
    },

    // additional properties, not important
    name: 'string',
}
```

I see you already have some tests in place for that, so maybe there are not that many new tests needed.

> > I think that your changes in MongoDB.prototype.isObjectIDProperty should be enough to fix coercion for all properties, so it should be enough to add few tests to verify
>
> That method is called only for ids at the moment. If we want to coerce all the properties, we'll have to refactor the code to loop through the properties, in many places. That's why I was wondering if we should do it in this PR.

Thanks for the explanation. I was naively assuming that `isObjectIDProperty` would be called for all properties.

I agree with you to leave support for ObjectID coercion in non-id properties out of scope of this pull request.